### ISSUE TITLE
Delete Widget Dynamically

### ISSUE THREAD

**Issue by lielran** [Shopify/dashing#611]
is it possible to delete widgets dynamically from a job ?

**Comment by tylermauthe** [comment]
Please direct support requests to Stack Overflow: stackoverflow.com/questions/tagged/dashing

**Comment by mapledyne** [comment]
Yes. I wrote a simple example job that can do just that here:
http://www.mapledyne.com/ideas/2015/6/30/delete-a-dashing-dashboard-widget

You basically just want to manipulate the Sinatra::Application.settings.history variable, but the code in that link should get you most of the way to where you want to be.

Or skip the post and go right to the gist file:
https://gist.github.com/mapledyne/6fb671c17c3f865309f3#file-delete-widget-rb

**Comment by tylermauthe** [comment]
Forcing a refresh, in my opinion, is a hack.

**Comment by mapledyne** [comment]
Forcing a refresh may be a hack, as what I do there (I'd agree on both counts, frankly), but what I describe above doesn't refresh the whole stack - in our situation, we only delete the widgets we're wanting deleted. We dynamically create dashboards that look to see what widgets exist to decide what to display. Having certain widget deleted, then, change the look of the dashboards as appropriate.

If there was an API call or similar to delete widgets individually, I'd love to hear it. Until then, manually removing them from the history variable has been very successful for us. 

In other words: do you have a non-hack solution to the problem of removing a widget? I'd love to learn.

**Comment by tylermauthe** [comment]
Well, I'm not sure that Dashing is really well suited for Dynamic Dashboards. Not without a lot of work.

However, a component itself could be dynamic. This can be achieved by the component JS manipulating the DOM to create and remove tiles which represent the data.

You have only one component, but it dynamically generates tiles for each data point.

I've been meaning to explore the possibility of creating a Meta-Component which comprehends lists of data-points compatible with other widgets. This would allow you to nest other components inside of the container and it would marshal the data automagically to update/create/delete sub-components. This should be possible, though it might not be as nice to use as I'm imagining.

**Comment by mapledyne** [comment]
We use it for dynamic dashboards lots, quite successfully, and I didn't find it to be a lot of work. It actually turned out to be much less than I expected. A job symlinks the appropriate dashboards to a template erb file (essentially one for every environment/set of VMs). This erb then looks up its filename which due to the symlink is the id of the environment. It looks up what VMs belong to that VM, then creates a row for each VM - in each row are all widgets with the appropriate VM id. 

This way we can create new environments, and dashing doesn't need to know about them (other than the job creating the new symlink automatically), and the template doesn't even need to know what widgets are present for a given VM, so it can report information that's only appropriate to itself.

We then deploy with puppet a script to each VM that in turn updates the widgets for the given VM.

Ultimately, QA builds a new set of VMs and within minutes there's a living dashboard for all of the systems in there, all without our intervention.

Hundreds of dashboards and thousands of widgets, and it works fantastically. Sure there's a hack here and there, but as we discover better ways, we can implement them pretty easily - all of those widgets and dashboards are ultimately a result of only three files, all reasonably small in size for the work they do for us.
